algorithms/src/polycommit/data_structures/powers.rs

```diff
@@ -0,0 +1,375 @@
+    #[derivative(Debug = "ignore")]
+    file: Arc<RwLock<File>>,
+    /// Group elements of the form `{ \beta^i \gamma G }`, where `i` ranges from 0 to `degree`.
+    /// These are used for hiding.
+    pub powers_of_beta_times_gamma_g: Arc<RwLock<BTreeMap<usize, E::G1Affine>>>,
+    /// The degree up to which we have powers.
+    /// Mostly just used for ensuring we download in the proper order,
+    /// length checks are actually done by checking file metadata.
+    degree: Arc<RwLock<usize>>,
+    _phantom_data: PhantomData<E>,
+}
+
+impl<E: PairingEngine> Default for PowersOfG<E> {
+    fn default() -> Self {
+        // This basically should never fail, hence the expect call.
+        Self::new(DEFAULT_PATH.clone()).expect("could not create default powers of G")
+    }
+}
+
+impl<E: PairingEngine> Clone for PowersOfG<E> {
+    fn clone(&self) -> Self {
+        Self::new(PathBuf::from(self.file_path.clone())).expect("could not clone powers of g")
+    }
+}
+
+impl<E: PairingEngine> ToBytes for PowersOfG<E> {
+    fn write_le<W: Write>(&self, mut writer: W) -> std::io::Result<()> {
+        // Serialize `powers_of_beta_times_gamma_g`.
+        (self.powers_of_beta_times_gamma_g.read().len() as u32).write_le(&mut writer)?;
+        for (key, power_of_gamma_g) in &*self.powers_of_beta_times_gamma_g.read() {
+            (*key as u32).write_le(&mut writer)?;
+            power_of_gamma_g.write_le(&mut writer)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<E: PairingEngine> FromBytes for PowersOfG<E> {
+    fn read_le<R: Read>(mut reader: R) -> std::io::Result<Self> {
+        // Deserialize `powers_of_beta_times_gamma_g`.
+        let mut powers_of_beta_times_gamma_g = BTreeMap::new();
+        let powers_of_gamma_g_num_elements: u32 = FromBytes::read_le(&mut reader)?;
+        for _ in 0..powers_of_gamma_g_num_elements {
+            let key: u32 = FromBytes::read_le(&mut reader)?;
+            let power_of_gamma_g: E::G1Affine = FromBytes::read_le(&mut reader)?;
+
+            powers_of_beta_times_gamma_g.insert(key as usize, power_of_gamma_g);
+        }
+
+        let powers = PowersOfG::default();
+        *powers.powers_of_beta_times_gamma_g.write() = powers_of_beta_times_gamma_g;
+        Ok(powers)
+    }
+}
+
+impl<E: PairingEngine> CanonicalSerialize for PowersOfG<E> {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), SerializationError> {
+        self.write_le(&mut *writer)?;
+        Ok(())
+    }
+
+    fn serialized_size(&self) -> usize {
+        todo!()
+    }
+}
+
+impl<E: PairingEngine> CanonicalDeserialize for PowersOfG<E> {
+    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
+        Ok(FromBytes::read_le(&mut *reader)?)
+    }
+}
+
+// NOTE: this drops the powers into a tmp file.
+// I assume this is only used for testing but this needs to be verified.
+impl<E: PairingEngine> From<Vec<E::G1Affine>> for PowersOfG<E> {
+    fn from(value: Vec<E::G1Affine>) -> Self {
```

Comment:
Could you clarify where this `from` method is used?

If we are confident it is exclusively for testing only, we should `cfg` guard it with `#[cfg(test)]` above the `impl` line (L153)

Reply:
I'm pretty confident it is, that's a really good idea :)

Reply:
Ah, it will actually prove quite tough, as the `setup` functions currently aren't `#[cfg(test)]` -- not sure if we want to guard those too?